File: python/recursions_backtrackings/day11/nQueens.py
def solve_n_queens(n):
    def solve(col):
        if col == n:
            ans.append(board.copy())
            return

        for row in range(n):
            if left_row[row] == 0 and \
               lower_diag[row + col] == 0 and \
               upper_diag[n - 1 + col - row] == 0:

                board[row] = board[row][:col] + "Q" + board[row][col+1:]

                left_row[row] = 1
                lower_diag[row + col] = 1
                upper_diag[n - 1 + col - row] = 1

                solve(col + 1)

                board[row] = board[row][:col] + "." + board[row][col+1:]

                left_row[row] = 0
                lower_diag[row + col] = 0
                upper_diag[n - 1 + col - row] = 0

    board = ["." * n for _ in range(n)]
    ans = []

    left_row = [0] * n
    lower_diag = [0] * (2 * n - 1)
    upper_diag = [0] * (2 * n - 1)

    solve(0)
    return ans


def print_chess_board(solution):
    n = len(solution)

    print("    ", end="")
    for i in range(n):
        print(f"  {chr(65+i)} ", end="")
    print()

    for i in range(n):
        print("    " + "+----" * n + "+")

        print(f"{i+1}   ", end="")
        for j in range(n):
            if solution[i][j] == 'Q':
                print("| ♛  ", end="")
            else:
                print("|    ", end="")
        print("|")

    print("    " + "+----" * n + "+")
    print("\n" + "=" * 50 + "\n")


# -------- Main --------
n = int(input("Enter board size (n): "))

solutions = solve_n_queens(n)

print(f"\nTotal Solutions: {len(solutions)}\n")

for sol in solutions:
    print_chess_board(sol)

File: java/recursions_backtrackings/Day11/NQueens.java
package recursions_backtrackings.Day11;
import java.util.*;

public class NQueens {
    public static void solve(int col, List<String> board, List<List<String>> ans, int[] leftRow, int[] upperDiagonal,
            int[] lowerDiagonal, int n) {
        if (col == n) {
            ans.add(new ArrayList<>(board));
            return;
        }

        for (int row = 0; row < n; row++) {
            if (leftRow[row] == 0 && lowerDiagonal[row + col] == 0 && upperDiagonal[n - 1 + col - row] == 0) {
                char[] currentRow = board.get(row).toCharArray();
                currentRow[col] = 'Q';
                board.set(row, new String(currentRow));

                leftRow[row] = 1;
                lowerDiagonal[row + col] = 1;

                upperDiagonal[n - 1 + col - row] = 1;

                solve(col + 1, board, ans, leftRow, upperDiagonal, lowerDiagonal, n);

                currentRow = board.get(row).toCharArray();

                currentRow[col] = '.';

                board.set(row, new String(currentRow));

                leftRow[row] = 0;
                lowerDiagonal[row + col] = 0;
                upperDiagonal[n - 1 + col - row] = 0;
            }
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter the size of board: ");
        int n = sc.nextInt();

        List<List<String>> ans = new ArrayList<>();
        List<String> board = new ArrayList<>();

        for (int i = 0; i < n; i++) {
            char[] row = new char[n];
            Arrays.fill(row, '.');
            board.add(new String(row));
        }

        int[] leftRow = new int[n];
        int[] upperDiagonal = new int[2 * n - 1];
        int[] lowerDiagonal = new int[2 * n - 1];

        solve(0, board, ans, leftRow, upperDiagonal, lowerDiagonal, n);

        System.out.println("Solutions in Chess Board Style:\n");

        for (List<String> solution : ans) {
            for (String row : solution) {
                for (int i = 0; i < row.length(); i++) {
                    if (row.charAt(i) == 'Q') {
                        System.out.print(" Q ");
                    } else {
                        System.out.print(" . ");
                    }
                }
                System.out.println();
            }
            System.out.println("---------------------");
        }

        sc.close();
    }
}
